crates/bevy_core/src/time/time.rs

```diff
@@ -34,58 +49,171 @@ impl Time {
     pub(crate) fn update_with_instant(&mut self, instant: Instant) {
         if let Some(last_update) = self.last_update {
             self.delta = instant - last_update;
-            self.delta_seconds_f64 = self.delta.as_secs_f64();
             self.delta_seconds = self.delta.as_secs_f32();
+            self.delta_seconds_f64 = self.delta.as_secs_f64();
+            self.time_since_startup += self.delta.mul_f32(self.relative_speed);
+        } else {
+            self.first_update = Some(instant);
+            self.time_since_startup += (instant - self.startup).mul_f32(self.relative_speed);
         }
 
-        let duration_since_startup = instant - self.startup;
-        self.seconds_since_startup = duration_since_startup.as_secs_f64();
         self.last_update = Some(instant);
     }
 
-    /// The delta between the current tick and last tick as a [`Duration`]
+    /// The delta between the current and last update, as a [`Duration`].
     #[inline]
     pub fn delta(&self) -> Duration {
         self.delta
     }
 
-    /// The delta between the current and last tick as [`f32`] seconds
+    /// The delta between the current and last update, as [`f32`] seconds
     #[inline]
     pub fn delta_seconds(&self) -> f32 {
         self.delta_seconds
     }
 
-    /// The delta between the current and last tick as [`f64`] seconds
+    /// The delta between the current and last update, as [`f64`] seconds.
     #[inline]
     pub fn delta_seconds_f64(&self) -> f64 {
         self.delta_seconds_f64
     }
 
-    /// The time since startup in seconds
+    /// The *virtual* delta between each fixed update step, as a [`Duration`].
+    /// Note that the true delta between fixed update steps varies.
```

Comment:
"virtual" and "true" sound confusing to me. One is not less "true" than the other. I'd rephrase to something like below, using "period" instead of "delta" to focus on the fixed-time simulation clock:
```suggestion
    /// The fixed update step period, as a [`Duration`].
    ///
    /// Fixed update steps are invoked with a fixed time monotonically increasing by
    /// this fixed delta time between each step. For example, at 20Hz, steps are simulated
    /// with fixed times 0s, 0.2s, 0.4s, 0.6s, ...
    ///
    /// Note that both the wall clock delta time between fixed update steps and the delta
    /// time between variable steps may vary, but the fixed timestep time remains a multiple
    /// of this delta time, and therefore the fixed simulation time will slowly drift compared to
    /// the variable frame time and wall clock time.
```

Reply:
"Period" is exactly the term to avoid. This is the biggest misunderstanding people have about fixed timesteps, see #2041. The period—as in the actual wall clock time that passes between fixed timesteps—is not something we can control. It depends on frame delta times.

All a fixed timestep does is guarantee that it runs a specific number of times for every second of frame time.

Hence, fixed delta is a pretend quantity for numerical stability and consistency.  We know the attached systems will run N times a second, so we use `Time::fixed_delta` to pretend that those steps are evenly spaced.

Reply:
Ok I've read through #2041 and your [comment at the end](https://github.com/bevyengine/bevy/issues/2041#issuecomment-898780701) is a good summary about the system ticking "at N steps per second" (I'd add "on average" even) and not "once every 1/N seconds". My suggestion is wrong in several aspects, most notably that we shouldn't talk about the _wall clock_ time at all because there's no way to measure it (I'll come back to it, there _do_ is drift, and `Instant::now()` does _not_ measure the wall clock time, but that' not important).

One thing I disagree with however is the usage of "period". Your argument is based on the assumption there's an implicit link between "period" and the wall clock time (or rather, `Instant::now()`). I don't think this is the case. A period in Physics in the inverse of the frequency, and is "the number of occurrences of a repeating event per unit of time" ([Wikipedia](https://en.wikipedia.org/wiki/Frequency)). There's no mention of any particular clock, because the period is measured between two instants, which themselves are measured relative to an implied clock. If the implied clock changes, the period may change. So although I agree we should be careful and document the fact there's no relation to `std::time::Instant` in particular, "period" is the most precise term to describe what that delta time between two fixed-step simulation ticks is.

Here's another proposal then:
```rust
/// The period of the fixed-step simulation clock.
///
/// Fixed-step simulation uses a monotonically increasing clock separate from the main
/// simulation clock ([`delta()`]). The time measured relative to that clock is called "fixed time".
/// Fixed update steps are invoked with a fixed time increasing by this fixed delta time between
/// each step. For example, at 20Hz (`fixed_delta` = 0.05), steps are simulated with fixed times
/// 0s, 0.05s, 0.1s, 0.15s, ...
///
/// Note that those fixed times are relative to the fixed-step clock, so the CPU time ([`Instant::now()`])
/// at which the update function is called may vary. The fixed-step simulation only guarantees
/// that its update runs 1/`fixed_delta` times per second _on average_, but not exactly once every
/// `fixed_delta` seconds. This is to favor the stability of the simulation itself over the regularity
/// at which the simulation runs, as fixed-step simulations are sensitive to the delta time between
/// the simulated steps, which should remain constant as much as possible. This is in contrast with
/// the main simulation clock, which attempts to be as regular as possible and tick every [`delta()`]
/// seconds relative to the CPU clock, most notably in order to sync with the monitor refresh rate
/// for rendering.
```

To be fair, most of this should be in the module documentation, which should introduce all those concepts in length, disambiguate between the various clocks, and explain how fixed-step simulation may tick zero or multiple times between the main simulation update.

Now, to clarify the "wall clock" term that I had the misfortune to introduce. This refers to the time in our physical world, and cannot be measured by any computer. In particular, `Instant::now()` is _not_ the wall clock time; instead this is generally the time of the internal CPU clock, which is affected by things like CPU frequency scaling, so is best referred to as the "CPU time" in my opinion. Quoting [the official documentation](https://doc.rust-lang.org/std/time/struct.Instant.html), it explicitly says this time can drift (emphasize mine):

> Note, however, that instants are not guaranteed to be steady. In other words, each tick of the underlying clock might not be the same length (e.g. **some seconds may be longer than others**). An instant may jump forwards or experience time dilation (slow down or speed up), but it will never go backwards.

The closest thing a computer has to a wall clock time is the OS time, which is NTP-synchronized, and _on average_ has no drift due to regular synchronization, but at any instant does have a drift (positive or negative).

So I propose to just avoid talking about "walk clock" time, and use "CPU time" instead to refer to `std::time::Instant`.

Reply:
Ok I've read through #2041 and your [comment at the end](https://github.com/bevyengine/bevy/issues/2041#issuecomment-898780701) is a good summary about the system ticking "at N steps per second" (I'd add "on average" even) and not "once every 1/N seconds". My suggestion is wrong in several aspects, most notably that we shouldn't talk about the _wall clock_ time at all because there's no way to measure it (I'll come back to it, there _do_ is drift, and `Instant::now()` does _not_ measure the wall clock time, but that' not important).

One thing I disagree with however is the usage of "period". Your argument is based on the assumption there's an implicit link between "period" and the wall clock time (or rather, `Instant::now()`). I don't think this is the case. A period in Physics in the inverse of the frequency, and is "the number of occurrences of a repeating event per unit of time" ([Wikipedia](https://en.wikipedia.org/wiki/Frequency)). There's no mention of any particular clock, because the period is measured between two instants, which themselves are measured relative to an implied clock. If the implied clock changes, the period may change. So although I agree we should be careful and document the fact there's no relation to `std::time::Instant` in particular, "period" is the most precise term to describe what that delta time between two fixed-step simulation ticks is.

Here's another proposal then:
```rust
/// The period of the fixed-step simulation clock.
///
/// Fixed-step simulation uses a monotonically increasing clock separate from the main
/// simulation clock ([`delta()`]). The time measured relative to that clock is called "fixed time".
/// Fixed update steps are invoked with a fixed time increasing by this fixed delta time between
/// each step. For example, at 20Hz (`fixed_delta` = 0.05), steps are simulated with fixed times
/// 0s, 0.05s, 0.1s, 0.15s, ...
///
/// Note that those fixed times are relative to the fixed-step clock, so the CPU time ([`Instant::now()`])
/// at which the update function is called may vary. The fixed-step simulation only guarantees
/// that its update runs 1/`fixed_delta` times per second _on average_, but not exactly once every
/// `fixed_delta` seconds. This is to favor the stability of the simulation itself over the regularity
/// at which the simulation runs, as fixed-step simulations are sensitive to the delta time between
/// the simulated steps, which should remain constant as much as possible. This is in contrast with
/// the main simulation clock, which attempts to be as regular as possible and tick every [`delta()`]
/// seconds relative to the CPU clock, most notably in order to sync with the monitor refresh rate
/// for rendering.
```

To be fair, most of this should be in the module documentation, which should introduce all those concepts in length, disambiguate between the various clocks, and explain how fixed-step simulation may tick zero or multiple times between the main simulation update.

Now, to clarify the "wall clock" term that I had the misfortune to introduce. This refers to the time in our physical world, and cannot be measured by any computer. In particular, `Instant::now()` is _not_ the wall clock time; instead this is generally the time of the internal CPU clock, which is affected by things like CPU frequency scaling, so is best referred to as the "CPU time" in my opinion. Quoting [the official documentation](https://doc.rust-lang.org/std/time/struct.Instant.html), it explicitly says this time can drift (emphasize mine):

> Note, however, that instants are not guaranteed to be steady. In other words, each tick of the underlying clock might not be the same length (e.g. **some seconds may be longer than others**). An instant may jump forwards or experience time dilation (slow down or speed up), but it will never go backwards.

The closest thing a computer has to a wall clock time is the OS time, which is NTP-synchronized, and _on average_ has no drift due to regular synchronization, but at any instant does have a drift (positive or negative).

So I propose to just avoid talking about "walk clock" time, and use "CPU time" instead to refer to `std::time::Instant`.

Reply:
You posted this comment twice.

Reply:
I know what period and frequency mean.

> Your argument is based on the assumption there's an implicit link between "period" and the wall clock time (or rather, Instant::now()). I don't think this is the case. 

No, I'm arguing that calling `time.fixed_delta()` a period is bad because it'll confuse people. Yes, _in the context of the simulation timeline_, the steps are spaced apart by `time.fixed_delta()`, but in actual CPU time they aren't. It's too easy for people to mix that up.

> CPU time != international atomic clock time

Agreed, but this is pretty irrelevant to my point that `time.fixed_delta()` is a "fake/pretend/virtual/simulated" quantity that decouples the attached systems from the frame rate, not a period you'll measure. I want the documentation to convey that.

Reply:
I see your point, but still think it's worth using period. I'd like others to chime in now, i don't think we'll convince each other. 😅

Reply:
Yeah, I'd like for other people to weigh in.

I think it's kinda funny how our opinions are flipped from the "steps per second" case.

> If you're talking about the fixed clock, then yes we can say **exactly N steps per second**, but that's not what I expect that expression to refer to for most people.

Replace "exactly N steps per second" with "period," and you have my perspective on this.

Reply:
I would agree that period is a weird word here. Nothing about a fixed delta simulation has anything to do with time periods.

Reply:
Gonna mark this conversation as resolved. I think the doc comments as of the latest commit give a concise warning (although the phrasing could probably be improved).
```rust
    /// The reciprocal of the rate set by [`Time::set_steps_per_second`], as [`f32`] seconds.
    ///
    /// A fixed timestep only guarantees that N steps run for each second of elapsed time, not that
    /// one step runs every [`Time::fixed_delta`] seconds. The exact CPU time between steps depends
    /// on the frame rate, so use [`Time::fixed_delta`] in systems for consistent behavior.
```
Feel free to re-open or make a new comment if you have other suggestions.